### PR TITLE
Unused coroutine return value analyzer

### LINKED ISSUE #12: Unused Coroutine Return Value Analyzer
#### Problem statement
Coroutines that are called directly are flagged by Visual Studio as having a return value that is unused. Calling a coroutine in this way will result in the coroutine's code not being executed.

#### Proposed solution
The call to the coroutine should be wrapped in StartCoroutine().

### PR DESCRIPTION
**Pull request by AnBucyk**
Fixes #12.

#### Checklist

- [X] I have read the [Contribution Guide](/CONTRIBUTING.md) ;
- [X] There is an approved issue describing the change when contributing a new analyzer or suppressor ;
- [X] I have added tests that prove my fix is effective or that my feature works ;
- [X] I have added necessary documentation (if appropriate) ;

#### Short description of what this resolves:
Adds an analyzer to detect and fix direct calls to coroutines, where their return value is unused. Such calls will not be executed.

#### Changes proposed in this pull request:
Provides a code fix that wraps the call to a coroutine in StartCoroutine().

### CODE REVIEW

File: src/Microsoft.Unity.Analyzers.Tests/UnityAnalyzerVerifier.cs [hunk 1]
```diff
@@ -37,12 +38,32 @@ public static Task VerifyAnalyzerAsync(string source, params DiagnosticResult[]
...
+
 			if (UnityPath.OnWindows())
-				installation = Path.Combine(installation, "Editor", "Data");
+				installationFullPath = Path.Combine(installation, "Editor", "Data");
 
-			var managed = Path.Combine(installation, "Managed");
-			yield return Path.Combine(managed, "UnityEditor.dll");
-			yield return Path.Combine(managed, "UnityEngine.dll");
+			// Unity installation might be within the Hub directory for Unity Hub installations
+			if (!Directory.Exists(installationFullPath))
```
Comment (AnBucyk) [inline src/Microsoft.Unity.Analyzers.Tests/UnityAnalyzerVerifier.cs]: Whoops. This is gonna fail on mac. Fix incoming.

File: src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
+using System;
+using System.Collections.Generic;
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs]: Unused namespace

File: src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+			title: Strings.UnusedCoroutineReturnValueDiagnosticTitle,
+			messageFormat: Strings.UnusedCoroutineReturnValueDiagnosticMessageFormat,
+			category: DiagnosticCategory.Correctness,
+			defaultSeverity: DiagnosticSeverity.Info,
+			isEnabledByDefault: true,
+			description: Strings.UnusedCoroutineReturnValueDiagnosticDescription);
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+		private static readonly Type coroutineReturnValue = typeof(System.Collections.IEnumerable);
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs]: Unused field

File: src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+			InvocationExpressionSyntax invocation, 
+			CancellationToken cancellationToken)
+		{
+			var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+			var newExpressionStatement = ExpressionStatement(
+				InvocationExpression(
+					IdentifierName("StartCoroutine"), 
+					ArgumentList(
+						SingletonSeparatedList<ArgumentSyntax>(
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs]: <ArgumentSyntax> will be inferred.
`
			var newExpressionStatement = ExpressionStatement(
				InvocationExpression(
					IdentifierName("StartCoroutine"), 
					ArgumentList(SingletonSeparatedList(Argument(invocation)))));
`

File: src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+			context.RegisterSyntaxNodeAction(AnalyzeInvocation, SyntaxKind.InvocationExpression);
+		}
+
+		private static void AnalyzeInvocation(SyntaxNodeAnalysisContext context)
+		{
+			var typeInfo = context.Compilation.GetTypeByMetadataName("System.Collections.IEnumerator");
```
Comment (sailro) [inline src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs]: nameof(System.Collections.IEnumerator) instead ?

Comment (jbevain) [reply src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs]: `typeof(System.Collections.IEnumerator).FullName`, nameof only returns the typename, not the fully qualified name.

Comment (sailro) [reply src/Microsoft.Unity.Analyzers/UnusedCoroutineReturnValue.cs]: > `typeof(System.Collections.IEnumerator).FullName`, nameof only returns the typename, not the fully qualified name.

Yep saw that after running tests :)

**Comment by sailro** [comment]
Apart from the minor comments, LGTM ! Thanks

**Pull request merged**